tests/src/test/java/alluxio/testutils/IntegrationTestUtils.java

```diff
@@ -12,44 +12,121 @@
 package alluxio.testutils;
 
 import static alluxio.util.network.NetworkAddressUtils.ServiceType;
+import static org.junit.Assert.assertEquals;
 
 import alluxio.AlluxioURI;
 import alluxio.ClientContext;
 import alluxio.Constants;
 import alluxio.client.file.FileSystem;
 import alluxio.client.file.FileSystemMasterClient;
+import alluxio.client.file.URIStatus;
+import alluxio.client.fs.FileOutStreamTestUtils;
 import alluxio.conf.PropertyKey;
 import alluxio.conf.ServerConfiguration;
-import alluxio.heartbeat.HeartbeatContext;
-import alluxio.heartbeat.HeartbeatScheduler;
+import alluxio.exception.status.UnavailableException;
+import alluxio.master.LocalAlluxioCluster;
 import alluxio.master.MasterClientContext;
 import alluxio.master.NoopMaster;
 import alluxio.master.PortRegistry;
+import alluxio.master.block.BlockMaster;
+import alluxio.master.file.meta.PersistenceState;
 import alluxio.master.journal.JournalUtils;
 import alluxio.master.journal.ufs.UfsJournal;
 import alluxio.master.journal.ufs.UfsJournalSnapshot;
 import alluxio.util.CommonUtils;
 import alluxio.util.FileSystemOptions;
 import alluxio.util.URIUtils;
 import alluxio.util.WaitForOptions;
-import alluxio.worker.block.BlockHeartbeatReporter;
+import alluxio.wire.WorkerInfo;
+import alluxio.worker.block.BlockMasterSync;
 import alluxio.worker.block.BlockWorker;
+import alluxio.worker.block.SpaceReserver;
 
 import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import org.junit.Assert;
 import org.powermock.reflect.Whitebox;
 
 import java.io.IOException;
 import java.net.URI;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Set;
 import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
 
 /**
  * Util methods for writing integration tests.
  */
 public final class IntegrationTestUtils {
 
+  /**
+   * - Checks that a file is in the {@link PersistenceState#TO_BE_PERSISTED} state.
+   * - Wait for the persist to occur
+   * - After occurring, makes sure that the URI status after the persist is
+   * {@link PersistenceState#PERSISTED}.
+   * - Checks that the file is in Alluxio and under storage.
+   *
+   * @param resource the local alluxio cluster resource
+   * @param path the path to check
+   * @param length expected size of the file
+   */
+  public static void checkPersistStateAndWaitForPersist(LocalAlluxioClusterResource resource,
+      FileSystem fs, AlluxioURI path, int length) throws Exception {
+    // check the file is completed but not persisted
+    URIStatus status = fs.getStatus(path);
+    assertEquals(PersistenceState.TO_BE_PERSISTED.toString(), status.getPersistenceState());
```

Comment:
is it sometimes possible that the persist is so fast, that the TO_BE_PERSISTED state is skipped?

Reply:
It is possible that is the case. I think this logic was wrong before this change though.